cmd/export_assets.go

```diff
@@ -0,0 +1,105 @@
+package cmd
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+	"github.com/stellar/stellar-etl/internal/input"
+	"github.com/stellar/stellar-etl/internal/transform"
+	"github.com/stellar/stellar-etl/internal/utils"
+)
+
+var assetsCmd = &cobra.Command{
+	Use:   "export_assets",
+	Short: "Exports the assets data over a specified range",
+	Long:  `Exports the assets that are created from payment operations over a specified ledger range`,
+	Run: func(cmd *cobra.Command, args []string) {
+		endNum, useStdout, strictExport := utils.MustCommonFlags(cmd.Flags(), cmdLogger)
+		startNum, path, limit := utils.MustArchiveFlags(cmd.Flags(), cmdLogger)
+
+		var outFile *os.File
+		if !useStdout {
+			outFile = mustOutFile(path)
+		}
+
+		paymentOps, err := input.GetPaymentOperations(startNum, endNum, limit)
+		if err != nil {
+			cmdLogger.Fatal("could not read assets: ", err)
+		}
+
+		seenIDs := map[uint64]bool{}
+		failures := 0
+		for _, transformInput := range paymentOps {
+			transformed, err := transform.TransformAsset(transformInput.Operation, transformInput.OperationIndex, transformInput.Transaction, transformInput.LedgerSeqNum)
+			if err != nil {
+				txIndex := transformInput.Transaction.Index
+				errMsg := fmt.Sprintf("could not extract asset from operation %d in transaction %d in ledger %d: ", transformInput.OperationIndex, txIndex, transformInput.LedgerSeqNum)
+				if strictExport {
+					cmdLogger.Fatal(errMsg, err)
+				} else {
+					cmdLogger.Warning(errMsg, err)
+					failures++
+					continue
+				}
+
+			}
+
+			// if we have seen the asset already, do not export it
+			if _, exists := seenIDs[transformed.AssetID]; exists {
```

Comment:
I'm a little wary of an in-memory map... Do we know that this will be accessible when ingesting from history archives and then when ingesting using captive-core? Might we invoke this command on different bounds of ledgers? We need total state of all assets for this check to work. Else, it definitely will have duplicates.

That said, I think this can be solved at the Airflow/BigQuery level through merges. Just flagging that at the ETL level, I'm not positive this works.

Reply:
In any case, might be good to add a comment here explaining why an in-memory map is an appropriate and sufficient internal data structure.

Reply:
You're right that overall we will have duplicate values. This in-memory map only ensures that there won't be duplicate assets within a given export. When we merge in Airflow, if an asset shows up multiple times in a single file, it will be inserted into the table multiple times. This means we need to ensure there is never duplicated info within a single export. From there, Airflow merges can handle the rest.

Reply:
Cool - this makes sense to me :)